### ISSUE TITLE
Implement ICECandidatePoolSize (gathering before SetLocalDescription/SetRemoteDescription)

### ISSUE THREAD

**Issue by om26er** [pion/webrtc#2892]
Here is the code
```
	// Prepare the configuration
	config := webrtc.Configuration{
		ICEServers:           r.iceServers,
		ICECandidatePoolSize: 10,
	}

	// Create a new RTCPeerConnection
	connection, err := webrtc.NewPeerConnection(config)
	if err != nil {
		panic(err)
	}

	connection.OnICECandidate(func(candidate *webrtc.ICECandidate) {
		log.Println(candidate)
	})
```

It doesn't do anything unless description is set

**Comment by om26er** [comment]
What I expect is the candidates gathering to start even before setting the description.

**Comment by nils-ohlmeier** [comment]
I'm not sure if Pion has a special mode for your expectation. But your expectation is certainly not true for browser implementations. The reason is that before you set any description the implementation doesn't know for how many streams it needs to provide ICE candidates and if bundle is going to be used or not.

I think the more reasonable expectation would be to say that with pools activated there should be a lot less delay in gathering candidates after setting a description, compared to without using an ICE candidate pool.

**Comment by Sean-Der** [comment]
This could be a great optimization! All for adding this @nils-ohlmeier @om26er 

To keep things simple we could treate ICECandidatePoolSize as a bool. If `!= 0` then gather before signaling.

Great first issue for someone to work on!

**Comment by nils-ohlmeier** [comment]
I didn't realize that Pion doesn't have support for ICE candidate pool yet.

Yes having the default pool size be 0 and meaning no gathering before set(Local|Remote)Description makes perfect sense to me. And then an API to set it to some positive number if pooling is desired.

**Comment by yeongin-ji** [comment]
@Sean-Der 
Hi! Since this was mentioned as a great first issue, I’d love to give it a try. 
Is anyone currently working on this?

**Comment by JoTurk** [comment]
@yeongin-ji no, no one is working on it as far as I know. it was on my personal todolist for 4.3 but i didn't start working on it, feel free to do so. 
I assigned you :)

**Comment by yeongin-ji** [comment]
Hello, I'd like to hear your thoughts regarding `ICECandidatePoolSize`.

The pooling logic currently implemented in [this PR](https://github.com/pion/webrtc/pull/3358) instructs the ICE Agent via the `ICEGatherer` to simply start gathering when the pool size is greater than 0, and then it pools all the triggered candidates. As @Sean-Der mentioned in this issue, this effectively behaves like a boolean toggle.

I agree with @JoTurk 's review comment that for this feature, the pool size needs to function as a more meaningful number.

However, I am confused about whether "Pool Size" here refers to the "number of ICE components" or the "number of ICE candidates."

[RFC 9429 Section 4.1.1](https://datatracker.ietf.org/doc/html/rfc9429#section-4.1.1) states:

> "If a size is specified for the ICE candidate pool, this indicates the number of ICE components to pre-gather candidates for."

And [Section 3.5.4](https://datatracker.ietf.org/doc/html/rfc9429#section-3.5.4) mentions:

> "... However, to accelerate cases where the application knows the number of ICE components to use ahead of time, it may ask the implementation to gather a pool of potential ICE candidates to help ensure rapid media setup."

On the other hand, looking at `selectCandidateMediaSection`, it selects only a single media section and assigns it to one `ICEGatherer`. This suggests that Pion operates on the premise of using only a single ICE component. If that is the case, the maximum meaningful pool size value would effectively be `1`.

If we interpret it as the number of candidates, the implementation might be relatively simple, but this seems to conflict with the standard.

I would appreciate your insights on how candidate pool size should be interpreted in this context.

I’ve only recently started studying WebRTC, so my understanding might be incorrect. If I’ve misunderstood anything, please feel free to correct me. 

Thank you! :)

**Comment by nils-ohlmeier** [comment]
@yeongin-ji your observations are correct. 🙂 
Pion up to now only supports a bundled ICE transport, which means that all media sections are sharing a single ICE transport or a single ICE component. 

However if you look at browser or other implementations you will see that they offer the bundling as an optional thing. Which means they can connect to endpoints which use one ICE component per media section.
That is also where the ICE connection pool originates from. If an endpoint needs to allocate ICE candidates for multiple media sections it makes sense to allocate them ahead of time to speed things up.

Since Pion only supports bundling an ICE connection pool isn't as useful, but I believe there are open tickets to add support for non bundled ICE transport, in which case an ICE connection pool would become very useful.

**Comment by JoTurk** [comment]
@yeongin-ji @nils-ohlmeier we're in the process of adding multiple ice write paths, for bonding and eventually we'll support non-bundled connections super soon (maybe in 4.3 or 4.4), so maybe we can return an error if pool is any value more than 1.

**Comment by yeongin-ji** [comment]
@nils-ohlmeier @JoTurk 
Thank you. That was a huge help in understanding the architecture.

In that case, before non-bundled connections are fully supported, I think the immediate steps I can take are as follows. Is this correct? Please let me know if there is anything else I should consider. :)

1. Add a pool size option to the ICE Agent.
2. Return an error if the pool size is greater than 1.
3. Provide an API to adjust the pool size.
